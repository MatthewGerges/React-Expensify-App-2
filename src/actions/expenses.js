import uuid from 'uuid'

//ADD_EXPENSE
//action generator

//={} allows you to destructure an empty object

export const addExpense = (
    {
        description = '', 
        note = '', 
        amount = 0, 
        createdAt = 0
    } = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        //called to generate random id
        description,
        note,
        amount,
        createdAt
    }
})

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

//REMOVE_EXPENSE
//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})