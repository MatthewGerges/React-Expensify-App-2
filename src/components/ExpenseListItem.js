import React from 'react';
//import { connect } from 'react-redux';
//import { removeExpense } from '../actions/expenses';
import {Link} from 'react-router-dom';

const ExpenseListItem = ({dispatch, description, amount, createdAt, id}) => (
    //destructuring props object
    <div>
        <Link to = {`/edit/${id}`}>
        <h3>{description}</h3>
        </Link>
        <p>{amount} = {createdAt}</p>
    </div>
)


//not props.dispatch just dispatch!!
//because Because the connect function intercepts the props coming from {...expense} and passes those along with dispatch into the component it wraps.

// store.dispatch(removeExpense({id: expenseOne.expense.id}));


// const MapStateToProps = (state) => {
//     return {
//         description: state.expenses.description,
//         amount: state.expenses.amount,
//         createdAt: state.expenses.createdAt
//     }
// }

// export default connect(MapStateToProps)(ExpenseListItem);

// export default ExpenseListItem;

//CONNECT ALLOWS YOU TO ACCESS DISPATCH PROP

export default ExpenseListItem;
//ExpenseListItem intercepts props and passes dispatch with it