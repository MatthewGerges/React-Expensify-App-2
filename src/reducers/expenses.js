
const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            //state.push(action.expense)
            //but push changes original state 
            //return state.concat(action.expense)
            return [
                ...state,
                action.expense
            ]
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id)
            //very important to destructure it as you filter/ return it
        case 'EDIT_EXPENSE':
            //state.map allows you to go through all the other items and change the one that matches
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else{
                    return expense;
                }
            })
        default: 
            return state;
    }
};

export default expensesReducer
//or at function declaration above: export default (state = expensesReducer...)