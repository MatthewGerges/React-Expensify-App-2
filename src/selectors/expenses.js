import moment from "moment";

//Get visible expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) =>{
        // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        //if type of startdate is not equal to a #, will always be true for non-numbers only if it is a number, filter expenses
        //startdate is 2 and created at is 1
        //if the above is false it will be false and will be filtered out
        //if expense 
        // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const createedAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createedAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createedAtMoment, 'day'): true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch
        //if any are false they will be removed from array and will not be visible expense
    //sort gets called on an array and returns an array
    }).sort((a, b) => {
        if (sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1  : -1;
            //if a.created At is not less than  b.createdAt, we will return -1 so a comes first
            //if a is not less it is -1 ad comes after?
        }
        else if (sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
            //if a is not less than b, -1 wil be returned and a will come first since it's more exp
        }
    })
};


export default getVisibleExpenses;

//returns filtered and sorted array