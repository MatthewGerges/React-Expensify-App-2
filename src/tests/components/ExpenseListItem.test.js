import React from 'react';
import {shallow} from 'enzyme';
//import {ExpenseListItem} from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';
//can grab default not named export

test('should render expense list item with a specific expense from expenses', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[1]} />)
    //... is the spread operator gives expenselist item access to all props - description, id, etx.
    expect(wrapper).toMatchSnapshot()
})