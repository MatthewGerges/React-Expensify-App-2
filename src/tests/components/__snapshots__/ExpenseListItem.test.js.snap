// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item with a specific expense from expenses 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      Rent
    </h3>
  </Link>
  <p>
    109500
     = 
    -345600000
  </p>
</div>
`;
